### PR TITLE
PHP 8.1: Added `enum` support to sniffs

### PR DESCRIPTION
**Pull request by kukulich**
Based on https://github.com/squizlabs/PHP_CodeSniffer/pull/3478

### CODE REVIEW

**Comment by kukulich** [comment]
Rebased on master and ready for review.

**Comment by gsherwood** [comment]
First, this is terrific. Thank you.

Curious as to any opinions on treating enums with the exact same sniff rules as classes, especially PSR2 and PSR12 standards.

I'm in 2 minds. Given they are basically classes, treating them in the same way makes sense to me. But the PSR standards obviously don't mention enums, so it's quite possible someone reports a bug that PHPCS is enforcing rules that are not stated in the PSR.

I think there is an argument to say they are implied in the PSR, given the line :
> The term "class" refers to all classes, interfaces, and traits.

and the contents of the RFC itself:
> Enumerations are built on top of classes and objects. That means, except where otherwise noted, “how would Enums behave in situation X” can be answered “the same as any other object instance.”

But keen for other opinons.

**Comment by jrfnl** [comment]
Just my two cents:
Treating enums like classes in every single way but the "signature" (which would need special handling) and the "case" statements (which should probably be treated similar to class constants), is the logical thing to do and is how I would expect sniffs to behave.

FYI: I'm hoping to review most of the work in this PR tomorrow (if still needed by then).

**Review by gsherwood — CHANGES_REQUESTED** [review]

File: src/Standards/Squiz/Tests/PHP/NonExecutableCodeUnitTest.2.inc [hunk 1]
```diff
@@ -45,6 +45,12 @@ trait Something {
 	}
 }
 
+trait Something {
```
[the remark this answers is not in the document]
Comment (kukulich) [reply src/Standards/Squiz/Tests/PHP/NonExecutableCodeUnitTest.2.inc]: Fixed.

**Review by gsherwood — APPROVED** [review]

**Comment by jrfnl** [comment]
@kukulich I've started reviewing - shall I pull a PR with various tweaks to your branch to get through the small stuff ?

**Comment by kukulich** [comment]
@jrfnl Feel free :)

**Comment by jrfnl** [comment]
PR https://github.com/kukulich/PHP_CodeSniffer/pull/1 has been pulled with some small improvements.

> Notes:
> * Most are just small test improvements to add additional safeguards.
> * The new test in `Squiz.Commenting.InlineComment` was breaking a test which had to be at the end of the file. Fixed by moving the new test up.
> * The new test in `Squiz.WhiteSpace.ControlStructureSpacing` wasn't actually testing anything. Fixed by wrapping it in a control structure.
> * The test improvements involving "valid name" sniffs will, in most cases, now cause the tests to fail. This is due to bug https://github.com/squizlabs/PHP_CodeSniffer/pull/3534, which was discovered during this review. The tests will pass once the upstream bugfix PR has been merged.
> * The `Squiz.Classes.ValidClassName` sniff needed an additional fix to be able to handle backed enums.

That concludes part 1 of the review (= review current fixes).

For part 2, I'm now going to look at an additional set of results from token searches to see if any of the additional sniffs flagged need tests/sniff fixes for enum support.

**Comment by jrfnl** [comment]
Not 100% done yet, but here are my preliminary findings:

Additional sniffs for which tests/support should be added (I've prepared the commits for a few already):
- [x] `PEAR.WhiteSpace.ScopeClosingBrace` - Needs tests.
- [x] `PSR12.Classes.OpeningBraceSpace` - Needs tests.
- [x] `Squiz.Commenting.ClosingDeclarationComment` - Needs support + tests
- [x] `Squiz.Commenting.DocCommentAlignment` - Needs support + tests. Probably will also need adjustment to support docblocks for enum cases.
- [x] `Squiz.WhiteSpace.ScopeClosingBrace` - Needs tests.
- [ ] `File::getMemberProperties()` - Should this get an exception for enums, along the same lines as for interfaces ?
- [ ] `File::findImplementedInterfaceNames()` - Should support finding the interfaces implemented by enums too
- [ ] `File::findEndOfStatement()` - Needs tests.
- [ ] `Generic.CodeAnalysis.UnnecessaryFinalModifier` - enums are final by design, so while the `final` modifier for methods is allowed, it is always unnecessary.
- [ ] `Generic.CodeAnalysis.UnusedFunctionParameter` - Needs tests, may also need to take enums implementing interfaces into account
- [ ] `Generic.Formatting.MultipleStatementAlignment` - Needs tests, also for case statements (once support has been added for those)
- [ ] `Generic.WhiteSpace.ScopeIndent` - Needs tests.
- [ ] `PSR12.Files.FileHeader` - Needs tests.
- [ ] `Squiz.Classes.SelfMemberReference` - Should actually probably support all OO tokens. Needs tests for all to be added.
- [ ] `Squiz.Commenting.ClassComment` - Should actually probably support all OO tokens. Needs tests for all to be added. Also take note of inconsistency with PEAR version of this sniff (which has previously already been adjusted to support all OO tokens).

Note sure about the `Squiz.Classes.ClassDeclaration` sniff. What do you think ?

**Comment by kukulich** [comment]
Rebased and green.

**Comment by jaroslavlibal** [comment]
I'm not sure if this is the right place to mention this, so apologies if not.

If I have an enum with some methods, like this:

```php
<?php declare(strict_types = 1);

namespace Model\Language;

enum Language: string
{

    case CS = 'cs';
    case EN = 'en';

    /** @return string[] */
        private static function getNames(): array
        {
            return [
            self::CS->value => 'cs_CZ',
            self::EN->value => 'en_US',
            ];
        }
}
```

Then firstly `phpcbf` breaks the formatting (in the example, the output after it has been "fixed") and secondly `phpcs` reports a problem on the sniff `PSR1.Files.SideEffects.FoundWithSymbols`.

> A file should declare new symbols (classes, functions, constants, etc.) and cause no other side effects, or it should execute logic with side effects, but should not do both.

**Review by faytekin — APPROVED** [review]

**Comment by rrajkomar** [comment]
Hello, 
Is there any ETA for this merge ? Currently php enums are unusable with PHPCS with is delaying any migration to PHP8.1 with Enums (at least without disabling analysis on those files).
Also is there any estimation on the 3.7.0 release date ?
Thanks.

**Comment by rrajkomar** [comment]
Hello, 
Is there any ETA for this merge ? Currently php enums are unusable with PHPCS with is delaying any migration to PHP8.1 with Enums (at least without disabling analysis on those files).
Also is there any estimation on the 3.7.0 release date ?
Thanks.

**Comment by gsherwood** [comment]
> Hello, Is there any ETA for this merge ? Also is there any estimation on the 3.7.0 release date ?
There is no ETA. I've been exceptionally busy and haven't been able to review the changes since I last approved.

**Comment by gsherwood** [comment]
@jrfnl @kukulich I'm keen to get a release out with at least some ENUM support.

I think what's here already is great. While more support could be added, I'd like to know if either of you see any blockers to getting this merged in. I know I'm not going to have time to work on any of the additional sniff changes, but I can at least get this merged and released to partial support is available.

**Comment by kukulich** [comment]
I would be happy with release with partial enum support :)

**Comment by jrfnl** [comment]
> @jrfnl @kukulich I'm keen to get a release out with at least some ENUM support.
> 
> I think what's here already is great. While more support could be added, I'd like to know if either of you see any blockers to getting this merged in. I know I'm not going to have time to work on any of the additional sniff changes, but I can at least get this merged and released to partial support is available.

Happy for this to be merged.

I could pull the extra changes which I worked on already either to this PR or separately, whichever is preferred.

Not all sniffs having support is IMO not a blocker for the release.

However, I do think it would be good if the below three from my previous comment got sorted before the release:

> - [ ] `File::getMemberProperties()` - Should this get an exception for enums, along the same lines as for interfaces ?
> - [ ] `File::findImplementedInterfaceNames()` - Should support finding the interfaces implemented by enums too
> - [ ] `File::findEndOfStatement()` - Needs tests.

* For the first I'd like a second opinion before actioning this.
* I can work on getting the second sorted.
* Anyone else up for looking at the third ?

Other than that for a release to go out, please be aware that IMO supporting intersection types will require the introduction of a new token (similar to what was done for union types), so that would mean that support for intersection types would need to wait until `3.8.0`.

**Comment by janedbal** [comment]
> Happy for this to be merged.

Great! We are really looking forward to this! Can we expect new version this week?

**Comment by acidjazz** [comment]
Almost 5 months since the release of PHP 8.1 Enums - lets merge and release!

**Comment by gsherwood** [comment]
This is now merged. Thanks all.

**Pull request merged**